File: can_ros2_practice/src/speed_monitor_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <can_msgs/msg/frame.hpp>

class SpeedMonitorNode : public rclcpp::Node
{
public:
    SpeedMonitorNode() : Node("speed_monitor_node")
    {
        subscription_ = this->create_subscription<can_msgs::msg::Frame>(
            "/from_can_bus",
            10,
            std::bind(&SpeedMonitorNode::on_can_frame, this, std::placeholders::_1)
        );

        RCLCPP_INFO(this->get_logger(), "speed_monitor_node started");
    }

private:
    void on_can_frame(const can_msgs::msg::Frame::SharedPtr msg)
    {
        if (msg->id != 0x100) {
            return;
        }

        if (msg->dlc < 1) {
            RCLCPP_WARN(this->get_logger(), "Received CAN ID 0x100, but data is empty");
            return;
        }

        int speed = static_cast<int>(msg->data[0]);

        if (speed >= 60) {
            RCLCPP_WARN(this->get_logger(), "Speed too high: %d km/h", speed);
        } else {
            RCLCPP_INFO(this->get_logger(), "Speed: %d km/h", speed);
        }
    }

    rclcpp::Subscription<can_msgs::msg::Frame>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SpeedMonitorNode>());
    rclcpp::shutdown();
    return 0;
}

File: 2026/06/2026-06-05_ros2_Speed-Alert-Node/ws/src/speed_watch_practice/speed_watch_practice/speed_monitor_node.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32


class SpeedMonitorNode(Node):
    def __init__(self):
        super().__init__('speed_monitor_node')
        self.declare_parameter('threshold', 60.0)
        self.threshold = self.get_parameter('threshold').value

        self.get_logger().info(
            f'Speed monitor started. threshold = {self.threshold} km/h'
        )

        self.subscription = self.create_subscription(
            Float32,
            '/vehicle/speed',
            self.speed_callback,
            10
        )

    def speed_callback(self, msg):
        speed = msg.data

        if speed >= self.threshold:
            self.get_logger().warn(
                f'Speed too high: {speed:.1f} km/h'
            )
        else:
            self.get_logger().info(
                f'Speed OK: {speed:.1f} km/h'
            )


def main(args=None):
    rclpy.init(args=args)

    node = SpeedMonitorNode()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()